### PR TITLE
Glioblastoma example notebook

### PR DESCRIPTION
**Pull request by njmei**
This PR is intended to address:
https://github.com/AllenInstitute/AllenSDK/issues/1648

Handy link to view the notebook itself:
https://app.reviewnb.com/AllenInstitute/open_dataset_tools/pull/3/

### CODE REVIEW

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb [hunk 1]
```diff
@@ -0,0 +1,4225 @@
...
+    "\n",
+    "The Ivy Glioblastoma Atlas Project (Ivy GAP) is a collection of data from glioblastoma brain tumors. The project is a collaboration between the Allen Institute for Brain Science and the Ben and Catherine Ivy Foundation. Glioblastoma is an aggressive brain cancer. Survival after diagnosis is just 12 to 15 months. An interactive atlas is available at https://glioblastoma.alleninstitute.org/ and consists of:\n",
+    "\n",
+    "\n",
+    "- Images of in situ hybridization experiments identifying where key genes are expressed in the tumor tissue (20x magnification)\n",
+    "- Matching histology images using Hematoxylin and Eosin (H&E) stain\n",
+    "- Gene expression masks for all ISH images\n",
+    "- Annotated mask images of tumor structures for all ISH and H&E images\n",
+    "- RNA sequencing data for 270 samples from 44 tumors\n",
+    "- Companion clinical database hosted at https://ivygap.swedish.org (registration required)\n",
```
Comment (wbwakeman) [inline Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb:26]: URL should be: https://ivygap.org/

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb [hunk 2]
```diff
@@ -0,0 +1,4225 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dcd3a346",
+   "metadata": {},
+   "source": [
+    "## Accessing Ivy Glioblastoma Atlas Project data with AWS S3\n",
+    "\n",
+    "The image data for the project is being made available as an Amazon Web Services (AWS) public dataset to enable computational scientists easy access to a rich, well-annotated data set for training and validation of machine learning, classification, and developing computer vision applications such as image segmentation. The data are free and publicly available from the AWS Simple Storage Service (S3) s3://allen-ivy-glioblastoma-atlas/ and/or https://console.aws.amazon.com/s3/buckets/allen-ivy-glioblastoma-atlas.\n",
```
Comment (wbwakeman) [inline Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb]: I'm still pretty confused by what access people should expect.  I don't think this console link works for non-admins, even if they are logged in with their AWS account.  Probably should remove that URL.  The S3 ARN should stay though.

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb [hunk 3]
```diff
@@ -0,0 +1,4225 @@
...
+    "\n",
+    "The root of the S3 bucket contains 3 metadata \\*.json files (`donor_metadata.json`, `specimen_metadata.json`, and `section_metadata.json`) that can be used to explore the dataset and determine download links. \n",
+    "\n",
+    "The relationship between donors, specimens, and sections is as follows:\n",
+    "- Each donor can have one or more (tumor tissue) specimens\n",
```
Comment (wbwakeman) [inline Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb:44]: I like this clear description.  One addition:
"Each donor (a human cancer patient) can have one or more (tumor tissue) specimens"

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb [hunk 4]
```diff
@@ -0,0 +1,4225 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "83fb5d06",
+   "metadata": {},
+   "source": [
+    "## Exploring section metadata and interacting with images\n",
+    "\n",
+    "The section metadata table includes information about sliced tissues sections obtained from specimens including primer sequence data used for in situ hybridization experiments.\n",
```
Comment (wbwakeman) [inline Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb:1413]: sliced tissues sections -> sliced tissue sections

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb [hunk 5]
```diff
@@ -0,0 +1,4225 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0a675543",
+   "metadata": {},
+   "source": [
+    "The section image table contains data as well as metadata for sub-images within a section. Images are loaded in as instances of the Python Image Library (PIL) `Image` class. For more details about this class, please see: https://pillow.readthedocs.io/en/stable/reference/Image.html\n",
+    "\n",
+    "For the purposes of this demo notebook, it is sufficient to know that one can use matplotlib plot some previews of `Image` objects:"
```
Comment (wbwakeman) [inline Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb:3359]: use matplotlib plot -> use matplotlib to plot

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb [hunk 6]
```diff
@@ -0,0 +1,4225 @@
...
+      "Downloading image from: s3://allen-ivy-glioblastoma-atlas/donor_15350/specimen_293591387/annotation/293591387_29_294218229_annotation.jpg\n",
+      "Downloading image from: s3://allen-ivy-glioblastoma-atlas/donor_15350/specimen_293591387/boundary/293591387_29_294218229_boundary.jpg\n",
+      "Downloading image from: s3://allen-ivy-glioblastoma-atlas/donor_15350/specimen_293591387/primary/293591387_32_294218211.jpg\n",
+      "Downloading image from: s3://allen-ivy-glioblastoma-atlas/donor_15350/specimen_293591387/annotation/293591387_32_294218211_annotation.jpg\n",
+      "Downloading image from: s3://allen-ivy-glioblastoma-atlas/donor_15350/specimen_293591387/boundary/293591387_32_294218211_boundary.jpg\n"
+     ]
+    }
+   ],
+   "source": [
+    "section_image_table = section_image_loader(\n",
```
Comment (wbwakeman) [inline Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb]: Oh man. I got a "MemoryError" while it was downloading the 10th image of the set. Not sure if there is anything to be done about it.  Unfortunately, it does prevent me from going forward with the notebook

**Review by danielsf — CHANGES_REQUESTED** [review]
What is the rationale for having the utility functions (`section_image_loader`, `section_metadata_cached_loader`, `section_image_dowloader`) defined in the notebook, rather than in a module that the notebook imports?

If we put the functions in a module then people who (like me) prefer working in scripts to working in notebooks can import the functions from their own scripts. I suspect this would be especially useful for someone who wants to start a job downloading the whole dataset and walk away from it while it runs.

I'm open to arguments why the notebook should remain as-is, but that is my opinion.

**Comment by djkapner** [comment]
I added come ReviewNB comments.
Why are these files in the root of this repo?

**Review by danielsf — APPROVED** [review]
Based on our conversation after stand up, this is ready to merge.

**Comment by njmei** [comment]
@djkapner Okay, I think I've addressed most of your comments. Let me know if anything else stands out.

**Review by djkapner — APPROVED** [review]

**Review by wbwakeman — COMMENTED** [review]
submitting these now.  Maybe they were not seen before?

File: Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb (hunk 6 above)
Comment (njmei) [reply Programmatic_access_of_Ivy_Glioblastoma_Atlas_Project_data.ipynb]: How much RAM does the system you tried this on have?

**Pull request merged**